File: problems/problems_2306/solution.ts
function distinctNames(ideas: string[]): number {
	const groups: Set<string>[] = Array.from({ length: 26 }, () => new Set());
	for (const s of ideas) {
		groups[s.charCodeAt(0) - 'a'.charCodeAt(0)].add(s.slice(1)); // 按照首字母分组
	}
	let ans: number = 0;
	for (let a: number = 1; a < 26; a++) { // 枚举所有组对
		for (let b: number = 0; b < a; b++) {
			let m: number = 0; // 交集的大小
			for (const s of groups[a]) {
				if (groups[b].has(s)) {
					m++;
				}
			}
			ans += (groups[a].size - m) * (groups[b].size - m);
		}
	}
	return ans * 2; // 乘 2 放到最后
};

export function Solve(inputJsonElement: string): any {
	const inputValues: string[] = inputJsonElement.split("\n");
	const ideas: string[] = JSON.parse(inputValues[0]);
	return distinctNames(ideas);
}

File: problems/problems_2306/solution.rs
use serde_json::{json, Value};

pub struct Solution;

use std::collections::HashSet;

impl Solution {
    pub fn distinct_names(ideas: Vec<String>) -> i64 {
        let mut groups = vec![HashSet::new(); 26];
        for s in ideas {
            groups[(s.as_bytes()[0] - b'a') as usize].insert(s[1..].to_string()); // 按照首字母分组
        }

        let mut ans = 0i64;
        for a in 1..26 { // 枚举所有组对
            for b in 0..a {
                let m = groups[a].iter().filter(|&s| groups[b].contains(s)).count(); // 交集的大小
                ans += (groups[a].len() - m) as i64 * (groups[b].len() - m) as i64;
            }
        }
        ans * 2 // 乘 2 放到最后
    }
}

#[cfg(feature = "solution_2306")]
pub fn solve(input_string: String) -> Value {
	let input_values: Vec<String> = input_string.split('\n').map(|x| x.to_string()).collect();
	let ideas: Vec<String> = serde_json::from_str(&input_values[0]).expect("Failed to parse input");
	json!(Solution::distinct_names(ideas))
}
